sklearn/metrics/_regression.py

```diff
@@ -1188,24 +1239,33 @@ def r2_score(
         return float("nan")
 
     if sample_weight is not None:
-        sample_weight = column_or_1d(sample_weight)
-        weight = sample_weight[:, np.newaxis]
+        sample_weight = column_or_1d(sample_weight, dtype=dtype)
+        weight = sample_weight[:, None]
     else:
-        weight = 1.0
+        weight = xp.asarray([1.0], dtype=y_true.dtype, device=device_)
```

Comment:
```suggestion
        weight = 1.0
```

should work, no?

Reply:
The concern here is esthetic rather than functional ? I thought it's nicer to explicit this way, because we explicit the dtype and device early, and do not rely on implicit conversion for later use. On the other hand it's more verbose.

Reply:
I prefer to use Python scalars in such expressions. It's more concise / easier to read.

Reply:
Any opinion on this @betatim / @glemaitre? See also the related: https://github.com/scikit-learn/scikit-learn/pull/27904#discussion_r1466098513.

Reply:
Looking at the original code I isn't immediately obvious to me why we are using it. As in, I think I'd need a comment to explain to me why to use `xp.array(...)` instead of `1.0`.

Reply:
I reverted back to `1` (in retrospect of this discussion I'll edit again for `1.0` :thinking: )